### DISCUSSION TITLE
Custom curved text class crops text

### DISCUSSION THREAD

**Discussion by timoostrich** [fabricjs/fabric.js#7799] [Q&A] [answered]
Hello everybody,

I really need some help on my custom curved text class.
Somehow my text gets cropped, when I convert itext to my custom class.

My text also gets cropped when I scale my object up (scaling down makes no problem).
As soon as the canvas gets refreshed it's not cropped anymore.

I added a screenshot shows the problem:
![crop-problem](https://user-images.githubusercontent.com/64577251/157830415-26fcdf2b-3673-40c4-bb1d-06e374d3b96f.jpg)

For me it looks like there's a caching problem (maybe because my custom function does something wrong regarding the rendering?).
I also looked at fabric.js source code but I really don't get it. Using objectCaching: false works sometimes (but breaks when my object has a stroke).

_I also tried to do it in another way (drawing directly to the ctx and not using a temp one to draw to the real canvas via drawimage later on). That worked, but than my drawings (circle guideline) are bigger than the text obj borders._

Here you can see a demo: https://codepen.io/timothy90/pen/eYZPjjv
In the demo that problem only occurs on the first conversion from itext to my custom curved text.

Any ideas how I can fix that? Or am I doing something completely wrong?

Thank you so much!
Timo

**Comment by timoostrich** [comment]
I just created another example, where I am drawing to the ctx directly instead of creating my own and using drawImage later on:
https://codepen.io/timothy90/pen/RwxbpQv

Is that the better way to use my custom curved text? I don't really not what would be some kind of best practice.
But, no matter which version is better: Here the object gets cropped, too, when objectCaching is not set to false.

**Comment by melchiar** [comment]
Feel free to check out my example of implementing curved text.

https://jsfiddle.net/melchiar/vhgtesoy/

**Reply by timoostrich** [reply]
Thanks for sharing but I really want to use a custom object class instead of using the new text on path feature.
I also need that knowledge for other custom objects I am using.

**Comment by timoostrich** [comment]
@asturur Could you have a quick look and point me in the right direction? You probably know what to do at first sight.
Thank you so much!

**Reply by asturur** [reply]
hi @timoostrich i look tomorrow! now is a bit late for me, i was digging into fabricJS email

**Reply by timoostrich** [reply]
@asturur Thanks, looking forward to it :)

**Comment by melchiar** [comment, marked as answer]
@timoostrich Without going too deeply into your code, issues of an incorrect cache size are usually caused by the object having the wrong dimensions when the cache is updated. You can force a cache update by calling `_updateCacheCanvas()`, but this needs to be done after the object's width and height properties are set. 

The reason the cache corrects itself after scaling the object in your example is that scaling triggers a new cache update, and by then the object has a width and height property.

https://codepen.io/melchiar/pen/BaJOWxm

**Reply by timoostrich** [reply]
Looks like that was the missing piece! I knew it's something like that but I did not know about that function.
My custom classes are working now, even with caching on.
Thank you so much!
